src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -204,6 +205,51 @@ protected Solution SetCurrentSolution(Solution solution)
             }
         }
 
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/> and updates <see cref="CurrentSolution"/> to the new value.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="oldSolution">Previous instance of the solution.</param>
+        /// <param name="newSolution">Transformed instance of the solution with updated <see cref="Solution.WorkspaceVersion"/>.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        protected bool SetCurrentSolution(Func<Solution, Solution> transformation, out Solution oldSolution, out Solution newSolution)
+        {
+            if (transformation == null)
+            {
+                throw new ArgumentNullException(nameof(transformation));
+            }
+
+            var currentSolution = Volatile.Read(ref _latestSolution);
+
+            while (true)
+            {
+                var transformedSolution = transformation(currentSolution);
+                if (transformedSolution == currentSolution)
+                {
+                    oldSolution = newSolution = currentSolution;
+                    return false;
+                }
+
+                newSolution = transformedSolution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
+                var replacedSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
+                if (replacedSolution == currentSolution)
+                {
+                    oldSolution = replacedSolution;
+                    return true;
+                }
+
+                currentSolution = replacedSolution;
+            }
```

Comment:
i'm honestly finding this really hard to read and grok.  Do we have to try to do this in a lock-free manner?  who is calling SetCurrentSolution so concurrently that we need this?

Reply:
The approach is the same as SetCurrentSolution(Solution solution) above.

---
In reply to: [389190889](https://github.com/dotnet/roslyn/pull/42228#discussion_r389190889) [](ancestors = 389190889)

Reply:
We can't be taking the lock while calling the transformation. So even using a lock it would not be a simple `lock(guard) { do stuff }` case.

---
In reply to: [389195023](https://github.com/dotnet/roslyn/pull/42228#discussion_r389195023) [](ancestors = 389195023,389190889)

Reply:
1. if it's the same, can we share some sort of single helper routine and have both entrypoints call into this?  i would hate to have a bug in one we finally fix, but not realize it affects the other.
2. even with teh same approach, i find it super difficult to reason about.  lock-free is not trivial and usually requires dozens of minutes of me sketching and trying to understand all the permutations and intricacies of what could happen here.

Reply:
> i it's the same, can we share some sort of single helper routine and have both entrypoints call into this?

I thought about it but such helper would be even more complicated.